File: lib/fetch_hf_models.go
package main

import (
	"encoding/json"
	"fmt"
	"io"
	"math"
	"net/http"
	"os"
	"regexp"
	"sort"
	"strconv"
	"strings"
	"sync"
	"time"
)

var (
	quantRE = regexp.MustCompile(`[Qq]\d[_.][Kk]?_?[A-Za-z0-9]*|IQ\d[_.][A-Za-z0-9]*`)
	shardRE = regexp.MustCompile(`-\d{5}-of-\d{5}`)
	paramRE = regexp.MustCompile(`(\d+(?:\.\d+)?)\s*[Bb](?:illion)?(?:\b|_)`)

	client = &http.Client{Timeout: 15 * time.Second}

	// Semaphore to limit concurrent requests
	sem = make(chan struct{}, 10)
)

// Model represents a model entry in the output JSON.
type Model struct {
	ID               string   `json:"id"`
	Name             string   `json:"name"`
	Repo             string   `json:"repo"`
	FilenameTemplate string   `json:"filename_template"`
	MinVRAMGB        float64  `json:"min_vram_gb"`
	Description      string   `json:"description"`
	Tags             []string `json:"tags"`
	downloads        int      // internal, not serialized
}

// Output is the top-level JSON structure.
type Output struct {
	Models []Model `json:"models"`
}

// apiModel represents the HuggingFace API search result.
type apiModel struct {
	ID        string `json:"id"`
	Downloads int    `json:"downloads"`
	Likes     int    `json:"likes"`
}

// apiModelDetail represents a detailed model response with siblings.
type apiModelDetail struct {
	Siblings []sibling `json:"siblings"`
}

type sibling struct {
	RFilename string `json:"rfilename"`
}

func estimateQ4GB(billions float64) float64 {
	return math.Round((billions*0.6+0.5)*10) / 10
}

func parseParamBillions(name string) float64 {
	m := paramRE.FindStringSubmatch(name)
	if len(m) < 2 {
		return 0
	}
	v, err := strconv.ParseFloat(m[1], 64)
	if err != nil {
		return 0
	}
	return v
}

func sizeTags(totalMemGB float64) []string {
	tags := []string{"1b", "3b"}
	if totalMemGB >= 8 {
		tags = append(tags, "7b", "8b")
	}
	if totalMemGB >= 16 {
		tags = append(tags, "14b")
	}
	if totalMemGB >= 32 {
		tags = append(tags, "30b", "32b", "34b")
	}
	if totalMemGB >= 64 {
		tags = append(tags, "70b", "72b")
	}
	return tags
}

func apiGet(url string) ([]byte, error) {
	sem <- struct{}{}
	defer func() { <-sem }()

	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
		return nil, err
	}
	req.Header.Set("User-Agent", "oi-cli/1.0")
	resp, err := client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	return io.ReadAll(resp.Body)
}

func deriveCliID(repoID string) string {
	parts := strings.Split(repoID, "/")
	repoName := parts[len(parts)-1]
	id := strings.ToLower(repoName)
	id = regexp.MustCompile(`-gguf$`).ReplaceAllString(id, "")
	id = regexp.MustCompile(`-instruct`).ReplaceAllString(id, "")
	id = regexp.MustCompile(`[._]`).ReplaceAllString(id, "-")
	id = regexp.MustCompile(`-+`).ReplaceAllString(id, "-")
	id = strings.Trim(id, "-")
	return id
}

func buildFilenameTemplate(filename string) string {
	loc := quantRE.FindStringIndex(filename)
	if loc == nil {
		return filename
	}
	return filename[:loc[0]] + "{quant}" + filename[loc[1]:]
}

func filterSharded(filename string) bool {
	return shardRE.MatchString(filename)
}

func formatNumber(n int) string {
	s := strconv.Itoa(n)
	if len(s) <= 3 {
		return s
	}
	var result []byte
	for i, c := range s {
		if i > 0 && (len(s)-i)%3 == 0 {
			result = append(result, ',')
		}
		result = append(result, byte(c))
	}
	return string(result)
}

func fetchModels(orgs []string, tags []string, maxParamB float64) []Model {
	type result struct {
		model Model
	}

	var wg sync.WaitGroup
	ch := make(chan result, 100)

	for _, org := range orgs {
		for _, size := range tags {
			wg.Add(1)
			go func(org, size string) {
				defer wg.Done()

				url := fmt.Sprintf(
					"https://huggingface.co/api/models?author=%s&search=%s+gguf&sort=downloads&direction=-1&limit=5&filter=text-generation",
					org, size,
				)

				data, err := apiGet(url)
				if err != nil {
					return
				}

				var results []apiModel
				if err := json.Unmarshal(data, &results); err != nil {
					return
				}

				for _, repo := range results {
					repoID := repo.ID
					downloads := repo.Downloads
					likes := repo.Likes

					paramB := parseParamBillions(repoID)
					if paramB > 0 && paramB > maxParamB {
						continue
					}

					detailData, err := apiGet(fmt.Sprintf("https://huggingface.co/api/models/%s", repoID))
					if err != nil {
						continue
					}

					var detail apiModelDetail
					if err := json.Unmarshal(detailData, &detail); err != nil {
						continue
					}

					// Collect single-file GGUFs only
					var ggufFiles []string
					for _, s := range detail.Siblings {
						fn := s.RFilename
						if !strings.HasSuffix(fn, ".gguf") || strings.HasPrefix(fn, ".") {
							continue
						}
						if filterSharded(fn) {
							continue
						}
						ggufFiles = append(ggufFiles, fn)
					}

					if len(ggufFiles) == 0 {
						continue
					}

					// Pick representative file (prefer Q4_K_M)
					rep := ""
					for _, fn := range ggufFiles {
						upper := strings.ToUpper(strings.ReplaceAll(fn, "-", "_"))
						if strings.Contains(upper, "Q4_K_M") {
							rep = fn
							break
						}
					}
					if rep == "" {
						rep = ggufFiles[0]
					}

					template := buildFilenameTemplate(rep)

					var minVRAM float64
					if paramB > 0 {
						minVRAM = estimateQ4GB(paramB)
					} else {
						minVRAM = 3.0
					}

					cliID := deriveCliID(repoID)

					repoName := repoID[strings.LastIndex(repoID, "/")+1:]
					modelName := strings.ReplaceAll(repoName, "-GGUF", "")
					modelName = strings.ReplaceAll(modelName, "-gguf", "")

					ch <- result{
						model: Model{
							ID:               cliID,
							Name:             modelName,
							Repo:             repoID,
							FilenameTemplate: template,
							MinVRAMGB:        minVRAM,
							Description:      fmt.Sprintf("%s downloads, %s likes on HuggingFace", formatNumber(downloads), formatNumber(likes)),
							Tags:             []string{"dynamic", strings.ToLower(org)},
							downloads:        downloads,
						},
					}
				}
			}(org, size)
		}
	}

	// Close channel when all goroutines complete
	go func() {
		wg.Wait()
		close(ch)
	}()

	// Deduplicate: keep highest downloads
	seen := make(map[string]Model)
	for r := range ch {
		m := r.model
		if existing, ok := seen[m.ID]; ok {
			if existing.downloads >= m.downloads {
				continue
			}
		}
		seen[m.ID] = m
	}

	// Sort by downloads descending
	models := make([]Model, 0, len(seen))
	for _, m := range seen {
		models = append(models, m)
	}
	sort.Slice(models, func(i, j int) bool {
		return models[i].downloads > models[j].downloads
	})

	return models
}

func main() {
	if len(os.Args) < 4 {
		fmt.Fprintf(os.Stderr, "Usage: %s <cache_path> <space_separated_orgs> <total_mem_gb>\n", os.Args[0])
		os.Exit(1)
	}

	cachePath := os.Args[1]
	orgs := strings.Fields(os.Args[2])
	totalMemGB, err := strconv.ParseFloat(os.Args[3], 64)
	if err != nil {
		totalMemGB = 8
	}

	tags := sizeTags(totalMemGB)
	maxParamB := (totalMemGB - 1) / 0.6

	models := fetchModels(orgs, tags, maxParamB)

	output := Output{Models: models}
	data, err := json.MarshalIndent(output, "", "  ")
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error encoding JSON: %v\n", err)
		os.Exit(1)
	}

	if err := os.WriteFile(cachePath, data, 0644); err != nil {
		fmt.Fprintf(os.Stderr, "Error writing cache file: %v\n", err)
		os.Exit(1)
	}

	fmt.Fprintf(os.Stderr, "Fetched %d models from HuggingFace\n", len(models))
}

File: lib/fetch_hf_models.py
import sys, json, re

try:
    from urllib.request import urlopen, Request
except ImportError:
    sys.exit(1)

cache_path = sys.argv[1]
orgs = sys.argv[2].split()
total_mem_gb = float(sys.argv[3] or "8")

API = "https://huggingface.co/api/models"
QUANT_RE = re.compile(r'[Qq]\d[_.][Kk]?_?[A-Za-z0-9]*|IQ\d[_.][A-Za-z0-9]*')
SHARD_RE = re.compile(r'-\d{5}-of-\d{5}')
PARAM_RE = re.compile(r'(\d+(?:\.\d+)?)\s*[Bb](?:illion)?(?:\b|_)')

def estimate_q4_gb(billions):
    """Estimate Q4_K_M file size in GB from param count in billions."""
    return round(billions * 0.6 + 0.5, 1)

def parse_param_billions(name):
    """Extract parameter count in billions from model name, e.g. '8B' -> 8.0"""
    m = PARAM_RE.search(name)
    if m:
        return float(m.group(1))
    return None

# Pick search size buckets based on total memory
# Rule of thumb: Q4_K_M ≈ 0.6 GB per billion params, plus ~1 GB overhead
size_tags = ["1b", "3b"]
if total_mem_gb >= 8:
    size_tags += ["7b", "8b"]
if total_mem_gb >= 16:
    size_tags += ["14b"]
if total_mem_gb >= 32:
    size_tags += ["30b", "32b", "34b"]
if total_mem_gb >= 64:
    size_tags += ["70b", "72b"]

# Max param count (billions) that fits in total memory at Q4_K_M
max_param_b = (total_mem_gb - 1) / 0.6

def api_get(url):
    req = Request(url, headers={"User-Agent": "oi-cli/1.0"})
    with urlopen(req, timeout=15) as resp:
        return json.loads(resp.read().decode())

seen = {}

for org in orgs:
    for size in size_tags:
        url = (f"{API}?author={org}&search={size}+gguf"
               f"&sort=downloads&direction=-1&limit=5&filter=text-generation")
        try:
            results = api_get(url)
        except Exception:
            continue

        for repo_info in results:
            repo_id = repo_info.get("id", "")
            downloads = repo_info.get("downloads", 0)
            likes = repo_info.get("likes", 0)

            # Estimate param count from repo name and check it fits
            param_b = parse_param_billions(repo_id)
            if param_b and param_b > max_param_b:
                continue

            try:
                details = api_get(f"{API}/{repo_id}")
            except Exception:
                continue

            siblings = details.get("siblings", [])

            # Collect single-file GGUFs only (skip sharded multi-part files)
            gguf_files = []
            for s in siblings:
                fn = s.get("rfilename", "")
                if not fn.endswith(".gguf") or fn.startswith("."):
                    continue
                if SHARD_RE.search(fn):
                    continue  # sharded — skip
                gguf_files.append(s)

            if not gguf_files:
                continue

            # Pick representative file (prefer Q4_K_M)
            rep = None
            for s in gguf_files:
                fn = s["rfilename"]
                if "Q4_K_M" in fn.upper().replace("-", "_"):
                    rep = fn
                    break
            if not rep:
                rep = gguf_files[0]["rfilename"]

            # Build filename template
            m = QUANT_RE.search(rep)
            if m:
                template = rep[:m.start()] + "{quant}" + rep[m.end():]
            else:
                template = rep

            # Estimate min_vram_gb from param count
            if param_b:
                min_vram = estimate_q4_gb(param_b)
            else:
                min_vram = 3.0

            # Derive short CLI id
            repo_name = repo_id.split("/")[-1]
            cli_id = repo_name.lower()
            cli_id = re.sub(r'-gguf$', '', cli_id)
            cli_id = re.sub(r'-instruct', '', cli_id)
            cli_id = re.sub(r'[._]', '-', cli_id)
            cli_id = re.sub(r'-+', '-', cli_id).strip('-')

            # Deduplicate: keep highest download count
            if cli_id in seen:
                if seen[cli_id]["_downloads"] >= downloads:
                    continue

            model_name = repo_name.replace("-GGUF", "").replace("-gguf", "")

            seen[cli_id] = {
                "id": cli_id,
                "name": model_name,
                "repo": repo_id,
                "filename_template": template,
                "min_vram_gb": min_vram,
                "description": f"{downloads:,} downloads, {likes:,} likes on HuggingFace",
                "tags": ["dynamic", org.lower()],
                "_downloads": downloads
            }

models = sorted(seen.values(), key=lambda m: m["_downloads"], reverse=True)
for m in models:
    del m["_downloads"]

output = {"models": models}
with open(cache_path, "w") as f:
    json.dump(output, f, indent=2)

print(f"Fetched {len(models)} models from HuggingFace", file=sys.stderr)
